### DISCUSSION TITLE
Context of company 'companyId'

### DISCUSSION THREAD

**Discussion by d1vak** [fattureincloud/api#224] [Q&A] [answered]
Hi, we are doing an integration with API v2 after the OAuth2 authorization we need to know a context of company 'companyId' in which we are logged. So we are trying to get list of all companies (with GET /user/companies) and filter them by accessToken and get ID of company, but it doesn't work that. For all companies accessToken's are 'null', is it possible to get this 'companyId' in another way?

**Comment by emajo** [comment, marked as answer]
Hi @divakpl, I don't understand what you mean by _filtering the companies by accessToken_, once you get a list of the companies with GET /user/companies, every company has a field `id`, that is in fact the companyId.

Does that solve your problem?
Emanuele.

**Reply by d1vak** [reply]
Hi @emajo, thank you very much for your answer. Probably I don't understand a FiC concept and architecture :(. At this moment our application is integrated with API v1, each of our users has individual 'api_id' and 'api_key' for his company, so when we fetching a data we getting only the data assigned to the specific company. In FiC v2 after OAuth2 authorization by user we can list all companies assigned to the user, but our application to integrate need a strict single company ID not a list. We don't know how to resolve this problem.

**Reply by valmoz** [reply]
Hi @divakpl 
the vast majority of the users will have only one company so you won't have this problem.
The only problem could be for accountants, who have access to a certain number of companies.

One way to select a company could be to propose to the user the list of available companies (retrieved using the [List User Companies](https://developers.fattureincloud.it/api-reference/#get-/user/companies) method) as a drop-down menu and then ask him to select the one he wants to interact with. This solution was already implemented by other integrations, so we know it works.

Thanks,
Mauro

**Reply by d1vak** [reply]
Hi @valmoz,

Thank you a lot for a quick feedback. So I understand that we need to send a extra data (after authorization) to our integration module for grabbing a correct company?. 

Now I wondering if an architecture of our application (we are using one client_id, client_secret per integration with provider) and model of user accounts are compatible wit FiC v2 architecture, because in our application each user account is equivalent of company. 

I think we need from FiC side something like a manager (accountants?) account for aggregation of all companies, and each company needs a its own credentials for the authorization.  I don't know it is possible?

**Reply by valmoz** [reply]
Hi @divakpl 
the API team's role is to help you understand the functionality of our APIs. If the V2 API is compatible with your application is up to you to decide.

What I can suggest is to start playing with our APIs, I think everything will get more clear once you start trying to use them for real.

My suggestion is to start creating a [Manual Token](https://developers.fattureincloud.it/docs/authentication/manual-authentication) on your Fatture in Cloud account (if you don't have one, you can create a normal account and require a free trial license for development purposes to the support), and then start trying our APIs using the [API Reference](https://developers.fattureincloud.it/api-reference#auth) or one of our [Quickstarts](https://developers.fattureincloud.it/docs/quickstarts/).

If you were using our V1 APIs previously, then the Manual Token is the most similar to the old auth method. the only difference is that once you obtain the token from the user you must perform the ListUserCompanies to retrieve the company id, then you can store it in your application. No additional data is required to start using our APIs.

How you select one of the IDs is up to you to decide. Other applications provide the user with the names of the companies returned from the ListUserCompanies as a drop-down menu, and let the user decide. Then they store the selected ID with the access token and start performing operations with our APIs. 

If, after trying our APIs with manual token, you still have some doubts about company ids, then the best solution is that you explain in a really detailed way your use case, so we can try to suggest you a possible solution. 

P.S.: we suggested the Manual Token because it's really easy to retrieve, but if you want you can use the OAuth methods in the future.

Thanks
Mauro

**Reply by valmoz** [reply]
Hi @divakpl were you able to make some progress on this issue?

**Reply by d1vak** [reply]
Hi, thanks a lot guys for your help. We have some conception how to resolve our problem with filtering and grabbing a correct companyId. Currently an implementation is inprogress. You can close discussion.

**Comment by sbenia75** [comment]
Buongiorno, 
sinceramente non ho capito come faccio a recuperare il codice "company_ID"

**Reply by valmoz** [reply]
Buongiorno @sbenia75
prima di aprire nuove domande vi pregherei di effettuare una ricerca per controllare se esistono delle guide o discussioni che parlano dello stesso argomento.

Il tema del company id è già trattato in questa guida: [Company-scoped methods](https://developers.fattureincloud.it/docs/basics/company-scoped-methods)
e ne abbiamo già parlato anche in questa discussione: #11 

Riassumendo: deve usare il metodo [List User Companies](https://developers.fattureincloud.it/api-reference/#get-/user/companies) e recuperare l'ID della company.

Mauro

**Reply by valmoz** [reply]
Buongiorno @sbenia75 è riuscito a recuperare il Company ID?
